src/timely-util/src/buffer.rs

```diff
@@ -81,11 +83,17 @@ where
             self.cap = Some(cap.delayed_for_output(cap.time(), self.port));
         }
         self.buffer.push(data);
-        if self.buffer.len() == self.buffer.capacity() {
+
+        // Limit, if possible, the lifetime of the allocations for data
+        // and consolidate smaller buffers if we're in the lucky case
+        // of a small domain for D
+        if self.buffer.len() >= 2 * self.previous_len {
```

Comment:
Trying to untangle this, this seems to ensure that the buffer doesn't grow beyond its initial capacity because we're flushing on length > capacity/2, which means that 2*previous length <= capacity. Is this reasoning correct?

Reply:
Seems correct to me! If we didn't flush, that means that the length is still <= half the capacity, which means we can double it at least once more without exceeding the capacity.

Reply:
Upon thinking about this, the tightest argument I could formulate was exactly what @teskje is saying. I added a code comment documenting the reasoning for why the two conditions work together to maintain an invariant across calls to `give`.